sqlalchemy_bigquery/_struct.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Mapping, Tuple
+
+import packaging.version
+import sqlalchemy.sql.default_comparator
+import sqlalchemy.sql.sqltypes
+import sqlalchemy.types
+
+from . import base
+
+sqlalchemy_1_4_or_more = packaging.version.parse(
+    sqlalchemy.__version__
+) >= packaging.version.parse("1.4")
+
+if sqlalchemy_1_4_or_more:
+    import sqlalchemy.sql.coercions
+    import sqlalchemy.sql.roles
+
+# We have to delay getting the type compiler, because of circular imports. :(
+type_compiler = None
+
+
+class STRUCT(sqlalchemy.sql.sqltypes.Indexable, sqlalchemy.types.UserDefinedType):
+    def __init__(
+        self,
+        *fields: Tuple[str, sqlalchemy.types.TypeEngine],
+        **kwfields: Mapping[str, sqlalchemy.types.TypeEngine],
+    ):
+        self.__fields = tuple(
+            (
+                name,
+                type_ if isinstance(type_, sqlalchemy.types.TypeEngine) else type_(),
+            )
+            for (name, type_) in (fields + tuple(kwfields.items()))
+        )
+
+        self.__byname = {name.lower(): type_ for (name, type_) in self.__fields}
```

Comment:
Let's check the BigQuery behavior when two fields differ only by case. If similarly named fields are allowed, we should keep the original casing, even if it would be nice to have case insensitivity most of the time.

Reply:
Looks like you're right.

```
CREATE TABLE my_dataset.struct_test
(
    struct_col STRUCT<Field INT64, field INT64>
)
```

Fails with "struct_col.field already exists"

Querying a table with different cases works fine:

```
CREATE TABLE my_dataset.struct_test
(
    struct_col STRUCT<field1 INT64, field2 STRING>
)
AS SELECT struct_col from UNNEST(
    [
        (STRUCT<struct_col STRUCT<field1 INT64, field2 STRING>>((123, "Hello"))),
        (STRUCT<struct_col STRUCT<field1 INT64, field2 STRING>>((456, "World")))
    ]
)
```

```
SELECT struct_col.Field1 as capped, struct_col.field1 as lowered FROM my_dataset.struct_test
```

Perhaps we'll hear from folks on https://issuetracker.google.com/199114438 what the desired behavior is, but this looks good to me, since they can't change from case insensitive to case sensitive without breaking existing queries.

Reply:
Oh look, case sensitivity is documented: https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#case_sensitivity

Reply:
Good find!